### PR TITLE
[925] Improve multiline dictionary and list indentation for sole function parameter

### PR DESCRIPTION
**Pull request by henriholopainen**
### Description

Making multiline lists and dictionaries less indented when passed as a function parameter and pairing parens with brackets/braces was discussed at in [#925](https://github.com/psf/black/issues/925). This PR introduces a conservative first step, where a multiline list or dict as the sole parameter to a function call will have special treatment for better readability. The feature is behind `hug_parens_with_braces_and_square_brackets` preview toggle.

### Checklist - did you ...

- [X] Add an entry in `CHANGES.md` if necessary?
- [X] Add / update tests if necessary?
- [X] Add new / update outdated documentation?

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -13,6 +13,7 @@
 ### Preview style
 
 - Fix merging implicit multiline strings that have inline comments (#3956)
+- Improve multiline dictionary and list indentation for sole function parameter (#3964)
```
Comment (JelleZijlstra) [inline CHANGES.md:16]:
```suggestion
- Multiline dictionaries and lists that are the sole argument to a function are now indented less (#3964)
```
The changelog entry should just describe the change; it's up to the reader to judge whether it's an improvement :)

File: docs/the_black_code_style/future_style.md [hunk 2]
```diff
@@ -113,6 +113,32 @@ my_dict = {
 }
 ```
 
+### Improved multiline dictionary and list indentation for sole function parameter
```
Comment (JelleZijlstra) [inline docs/the_black_code_style/future_style.md:116]: Thanks for adding this documentation!

File: tests/data/cases/preview_hug_parens_with_braces_and_square_brackets.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
+# flags: --preview
+def foo_brackets(request):
+    return JsonResponse(
```
Comment (JelleZijlstra) [inline tests/data/cases/preview_hug_parens_with_braces_and_square_brackets.py:3]: Ideas for additional tests:

- Test that comments in various places are preserved (e.g., after the opening `(` or before the closing `)`)
- Test where the argument is a list comprehension
- Test where the argument is a set literal

**Review by JelleZijlstra — APPROVED** [review]

**Comment by JelleZijlstra** [comment]
@henriholopainen seems like there's a crash on some code in Hypothesis (from the diff-shades failure). Can you take a look?

**Comment by henriholopainen** [comment]
Located the issue in Hypothesis (and other places); opening and closing brackets/braces don't necessary belong to the same entity. For example:

python```
foooooooooooooooooooo([{c: n + 1 for c in range(256)} for n in range(100)] + [{}], {size})
```

becomes currently:

python```
foooooooooooooooooooo([
    {c: n + 1 for c in range(256)} for n in range(100)] + [{}], {size
})
```

which is clearly very wrong. I'll add some test cases and look for a sophisticated approach for a fix tomorrow.

**Comment by henriholopainen** [comment]
Test cases added and logic fixed

File: CHANGES.md [hunk 4]
```diff
@@ -12,7 +12,10 @@
 
 ### Preview style
 
-<!-- Changes that affect Black's preview style -->
+- Fix merging implicit multiline strings that have inline comments (#3956)
+- Multiline dictionaries and lists that are the sole argument to a function are now
+  indented less (#3964)
+- Allow empty first line after block open before a comment or compound statement (#3967)
```
Comment (JelleZijlstra) [inline CHANGES.md:18]:
```suggestion
- Multiline dictionaries and lists that are the sole argument to a function are now
  indented less (#3964)
```

The other two were released

File: src/black/linegen.py [hunk 5]
```diff
@@ -815,6 +815,19 @@ def _first_right_hand_split(
     tail_leaves.reverse()
     body_leaves.reverse()
     head_leaves.reverse()
+
+    if Preview.hug_parens_with_braces_and_square_brackets in line.mode:
+        if (
+            tail_leaves[0].type == token.RPAR
+            and tail_leaves[0].value
+            and tail_leaves[0].opening_bracket is head_leaves[-1]
```
Comment (JelleZijlstra) [inline src/black/linegen.py:823]:
```suggestion
            and tail_leaves[0].opening_bracket == head_leaves[-1]
```
I don't think we can guarantee identity

File: src/black/linegen.py [hunk 6]
```diff
@@ -815,6 +815,19 @@ def _first_right_hand_split(
...
+            and body_leaves[-1].type in [token.RBRACE, token.RSQB]
+            and body_leaves[-1].opening_bracket is body_leaves[0]
```
Comment (JelleZijlstra) [inline src/black/linegen.py:825]:
```suggestion
            and body_leaves[-1].opening_bracket == body_leaves[0]
```

File: CHANGES.md (hunk 4 above)
Comment (henriholopainen) [reply CHANGES.md:18]: Sorry, bad merge on this

File: src/black/linegen.py (hunk 6 above)
Comment (henriholopainen) [reply src/black/linegen.py:825]: If we don't check for identity here, cases like

```
baaaaaaaaaaaaar([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], {x}, "a string", [1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
```

will format to

```
baaaaaaaaaaaaar([
    1, 2, 3, 4, 5, 6, 7, 8, 9, 10], {x}, "a string", [1, 2, 3, 4, 5, 6, 7, 8, 9, 10
])
```

which is not something we want.

File: src/black/linegen.py (hunk 5 above)
Comment (henriholopainen) [reply src/black/linegen.py:823]: Why not? The whole `omit` system in `_first_right_hand_split` seems to rely on `id(leaf)`, so the leaf objects should be stable and thus identity comparisons should be safe also here. Or am I missing something?

Comment (JelleZijlstra) [reply src/black/linegen.py:823]: Thanks, I was confused

**Comment by hauntsaninja** [comment]
Thanks for the PR! While I like the style, it's a fairly big change for black and goes against the philosophy of diff reduction (in the case you add a new argument). I wonder therefore if it makes sense to have magic trailing commas after the single arg prevent this reformatting.

**Comment by henriholopainen** [comment]
> Thanks for the PR! While I like the style, it's a fairly substantive change for black and goes against the philosophy of diff reduction (in the case you add or remove arguments). I wonder therefore if it makes sense to have magic trailing commas after the single arg prevent this reformatting.

Do you mean that we should instead format

```
foo({0: "looooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooong line"},)
```

as it would be formatted without the trailing comma:

```
foo({
    0: "looooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooong line"
})
```

?

**Comment by henriholopainen** [comment]
> ```python
> foo(*[
> 	pretend,
> 	many,
> 	and,
> 	long,
> 	arguments,
> 	# ... ..................................................................
> 	*another_list
> ])
> ```

Currently it doesn't, but adding support for that would make sense IMO.

**Comment by hauntsaninja** [comment]
It looks like Black already does what I suggested, I went ahead and added a test case and docs for it here: https://github.com/psf/black/pull/3991

**Comment by JelleZijlstra** [comment]
Maybe I was too hasty in merging this. I think this makes things generally look better, but we should definitely think more about whether we want to go through before we move this change to stable.

**Comment by charliermarsh** [comment]
One question, as both a user and a Ruff maintainer: is it intended that this also applies to parenthesized lists, sets, etc., outside of function calls? For example, with this commit:

```python
# Input
([1, 2, 3,])

# HEAD^
(
    [
        1,
        2,
        3,
    ]
)

# HEAD
([
    1,
    2,
    3,
])
```

**Comment by cthoyt** [comment]
This makes me so very happy, thanks @JelleZijlstra :)

**Pull request merged**